Add snippets.symbol_visibility_header() method

xclaesse opened the pull request:
Defining public API in a cross platform library is painful, especially
on Windows. Since every library have to define pretty much the same
macros, better do it in Meson.

xclaesse commented:
Cases I'm unsure: clang-cl? cygwin? msys2 gcc?

dcbaker commented:
Question, why not put this in the windows module?

xclaesse commented:
> Question, why not put this in the windows module?

I did that initially, then saw that we have gnu visibility to deal with too. The pain is not limited to Windows.

tristan957 commented:
@dcbaker this also defines the GCC/clang attributes.

jpakkane commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -586,3 +586,68 @@ methods:
+
+    If this header is installed it should be included at the beginning of every
+    public headers. If it is not being installed, it should be included at the
+    beginning of every source files, and it is the responsability of every
```
After "every" you need to use singular rather than plural case. So "every public header", "every source file" and "every application".

xclaesse replied:
Thanks, fixed.

xclaesse commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -586,3 +586,68 @@ methods:
+      type: str
+      description: Header file name.
+  kwargs:
+    api_variable:
```
I did not like _variable neither, so I just removed it. It's now `api`, `compilation` and `static_compilation`. Short and simple IMHO.

dcbaker commented:
Okay, then this should really live in the clike compiler mixin not in the base compiler

xclaesse commented:
> This is such a notable feature that it should get a mention in the proper docs as well, not only in the reference part.

Added a mention in `gnu_symbol_visibility` kwarg doc, do you have other places in mind?

> Okay, then this should really live in the clike compiler mixin not in the base compiler

I wasn't sure it's worth moving implementation into Compiler class, but since it needs different code for at least msvc and gnu, that would make sense indeed.

tristan957 commented:
Thinking about naming consistency...I added `java.generate_native_headers()`. Wondering if I should have named it `java.native_headers()` instead similar to how this PR is `cc.symbol_visibility_header()`. No verb. Naming is hard :(

xclaesse commented:
Added `static_only` kwarg, when set to true _STATIC_COMPILATION gets defined in the header, so no need to define it manually in cflags and pkgconfig file any more. That's how glib does it.

Also I fixed the doc, installing the generated header is always required. I was wrong to think it was a project choice.

xclaesse commented:
> No verb. Naming is hard :(

I think most of the time we indeed avoid verbs in Meson API, there are very few get_foo(). But that's not entirely consistent.

tristan957 commented:
Oh well, can always be fixed in a future release.

eli-schwartz commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+                #endif /* {static_compilation} */
+                ''')
+
+        if self.compiler.info.is_windows():
```
This is something I found really annoying about the cmake version, because you generate a file that isn't comprehensibly useful in the general case. What's the point in saving a few bytes for cross-platform defines?

xclaesse replied:
I'm not sure it is possible to have a static file that works in all cases tbh. How do you detect you're building a DLL, regardless of the compiler, from within a macro?

eli-schwartz replied:
This review comment is responding to the fact that the file contents are generated by checking in meson code whether you're building for Windows or `GNUC`, instead of generating both everywhere and using preprocessor checks to determine which one is used.

Generating both everywhere and using preprocessor checks to determine which one is used, results in a header that serves as a good teaching example and can be extracted from any OS and embedded into a third-party derivation.

I have been previously annoyed that cmake generates Windows-specific snippets, which means that the example file it generates -- and which people try to use in PRs to the WrapDB -- cannot be reused in a Meson port.

xclaesse replied:
Ah, I see, makes more sense, I'll update.

xclaesse replied:
I changed the code to have a common block for all platforms, but Meson still have to define `MESON_HOST_IS_WINDOWS` or `MESON_HAS_GNU_SYMBOL_VISIBILITY`. I don't think the latter can be determined by preprocessor, can it? The former could be replaced by something like `#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__NT__)` but I'm actually not entirely sure what's the right case where we actually need dllexport/dllimport, clang-cl, cygwin?

xclaesse replied:
What, actually there is an example there: https://gcc.gnu.org/wiki/Visibility.

xclaesse replied:
Ok updated again, now without any check at meson level. Is that what you had in mind?

eli-schwartz commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+                ''')
+        elif self._has_argument_impl('-fvisibility=hidden'):
+            content += textwrap.dedent(f'''\
+                #ifndef {api}
+                #  ifdef {compilation}
+                #    define {api} __attribute__((visibility("default")))
+                #  else /* !{compilation} */
+                #    define {api}
+                #  endif /* !{compilation} */
+                #endif /* {api} */
```
This is incorrect, because if the compiler supports visibility=hidden then there is *never* any circumstance, period, end of story, in which it's bad or wrong to define it.

You do not need to leave it empty when using it as a public header for linking against. You SHOULD NOT leave it empty, because it is confusing to change cases around. Also, it perpetuates the erroneous belief that you need to do so, which means Meson is teaching bad behavior. :)

xclaesse replied:
That's entirely subjective to judge if it's bad or not. You don't *need* to define the attribute, but you *can* define it if you want. Both are just equivalently good IMHO.

I copied this from how glib does it, but I don't mind changing it to always define the attribute if it's confirmed that it has no impact.

eli-schwartz replied:
No, it's objectively bad for exactly the social reasons I explained.

xclaesse replied:
"social reasons" that's by definition subjective :D

xclaesse replied:
Bah, changed to always set the attribute, that's what cmake does too.

eli-schwartz reviewed (commented):
This is essentially cmake's `GENERATE_EXPORT_HEADER()`and incidentally shares my frustration with cmake's implementation.

The block of code in question can be statically determined in a cross-platform manner and checked into git, so generating it at build time seems silly... but worse, it teaches bad behavior to people who would check this sort of thing into git.

And the test suite even bends over backwards in order to avoid the real problem here, which is that MSVC doesn't support default_library=both because you cannot define the API as two different things entirely.

xclaesse commented on docs/yaml/objects/compiler.yaml:
````diff
@@ -586,3 +586,95 @@ methods:
+    `mylib/meson.build`:
+    ```meson
+    default_library = get_option('default_library')
+    if host_machine.system() == 'windows' and default_library == 'both'
````
I think it is a DLL think whatever is the compiler, but could be wrong.

xclaesse replied:
Doc suggests that's indeed needed for GNU compilers too: https://gcc.gnu.org/wiki/Visibility. I replaced that Meson check by preprocessor check now anyway.

xclaesse commented:
> This is essentially cmake's `GENERATE_EXPORT_HEADER()`

I did not know cmake has this, do you have an example of project that uses that? I'm currious of what it contains.

eli-schwartz commented:
https://github.com/mesonbuild/wrapdb/pull/309

xclaesse commented:
> [mesonbuild/wrapdb#309](https://github.com/mesonbuild/wrapdb/pull/309)

Lol, cmake does it even weirder, they define it twice the same way...

```
#    ifdef libhighs_EXPORTS
        /* We are building this library */
#      define LIBHIGHS_EXPORT __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define LIBHIGHS_EXPORT __attribute__((visibility("default")))
#    endif
```

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
```
use kwargs.get() with a fallback instead

xclaesse replied:
No, that's a trap of typed_kwargs, it sets None as default value.

tristan957 replied:
Thanks for the lesson :)

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+                #  endif /* !{static_compilation} */
+                #endif /* {api} */
+                ''')
+        elif self._has_argument_impl('-fvisibility=hidden'):
```
I think this check is better written as callling the implementation of `compiler.has_function_attribute('visibility:hidden')`

xclaesse replied:
I replaced that check with `__GNUC__ >= 4` proprocessor, which is what https://gcc.gnu.org/wiki/Visibility suggests.

eli-schwartz replied:
IIRC we discussed this in the wrapdb at one point @xclaesse because a wrap you wrote turned out to be broken with has_argument, but worked with has_function_attribute.

eli-schwartz replied:
Anyway, the `__GNUC__` preprocessor check is accurate as it declares whether the compiler supports the GNU C dialect including visibility attributes, so no worries there.

xclaesse commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
```
Fixed, I prepend an underscore in that case.

tristan957 replied:
Aren't identifiers starting with an `_` reserved for the C implementation?

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -743,3 +747,59 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+            #        define {api} extern __declspec(dllimport)
+            #      endif /* !{compilation} */
+            #    endif /* !{static_compilation} */
+            #  elif __GNUC__ >= 4
```
Feels like a trap to me when all compilers supported by Meson can tell you if they support this or not, but if others don't agree then it can just be fixed in a future bug report.

xclaesse replied:
That's what says GNU's own documentation...

xclaesse commented:
> Okay, then this should really live in the clike compiler mixin not in the base compiler

@dcbaker done now.

xclaesse commented:
Finally passed the CI, ready for review.

tristan957 commented:
I like the idea of this function existing in a `clike` module rather than on the compiler itself.

xclaesse commented:
Anyone for a review please?

jpakkane commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -586,3 +586,95 @@ methods:
+    of a library. Depending on the type of compiler and the platform, this will
+    typically use `__declspec(dllexport)`, `__declspec(dllimport)` or
+    `__attribute__((visibility("default")))`. It is only supported for C and C++
+    compilers.
```
Instead of producing different files for different compilers, could we not make it always do the same thing with the appropriate ifdefs? It's not terribly complicated after all. Wit that change this would not need to live inside the compiler object, which is a slightly unintuitive place for it.

eli-schwartz replied:
I said the same thing in my review above. @xclaesse acceded to my suggestion.

The actual implementation in `compiler.py` now writes the same file everywhere, after taking into account the fact that it inserts the string values of the method kwargs into that file.

Effectively, it is now `configure_file()` with a builtin template.

P.S. 

> It's not terribly complicated after all.

This is also why I believe it's best described in a FAQ rather than implemented as a Meson function. My personal feelings on the matter are that users should never use this -- it does pointless file I/O at build time for something that could be checked into git.

But apparently people really want to "have meson guarantee it is correct", for the same reason that when using cmake they want cmake to generate a similar file "to have cmake guarantee it is correct".

xclaesse replied:
Yes, everyone gets it wrong, today I discovered that GLib got it wrong forever too. And it seems that even this PR got it wrong according to conversations I had with various people on the subject.

xclaesse replied:
I now integrated the changes we made in glib since then, the macro is slightly different.

bugaevc commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -586,3 +586,95 @@ methods:
+        If set to true, '<PROJECT_NAME>_STATIC_COMPILATION' is defined inside the
+        generated header.
+  description: |
+    Generate a header file that define a macro to be used to mark all public APIs
```
```suggestion
    Generate a header file that defines a macro to be used to mark all public APIs
```

xclaesse replied:
fixed

bugaevc commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -586,3 +586,95 @@ methods:
+
+    The generated header file should be installed using `install_headers()`.
+
+    `static_only: true` only has an affect when building for Windows platform.
```
```suggestion
    `static_only: true` only has an effect when building for Windows platform.
```

xclaesse replied:
fixed

bugaevc commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -734,3 +743,57 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '0.63.0')
+    def symbol_visibility_header_method(self, args: T.Tuple[str], kwargs: 'SymbolVisibilityHeaderKW') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp'}:
```
Will this also work for Objective-C / Objective-C++?

xclaesse replied:
I don't know how it works with objc, but if someone who knows can confirm it works, it can be added there, yes.

ePirat replied:
It works the same, Objective-C is a superset of C, so all you can do there, you can in Objective-C too.

xclaesse replied:
Ok, updated.

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
+                #endif /* {static_compilation} */
+                ''')
+        content += textwrap.dedent(f'''
+            #if (defined(_WIN32) || defined(__CYGWIN__)) && !defined({static_compilation})
```
So I would like to merge this given every project that ever exists need to do a configure_file() to get it right given the existence of static_compilation and compilation in this format string.

tristan957 replied:
Hmm I guess I am an idiot, and these would be constants and not require configure_file() :D

xclaesse replied:
Right, you can typically commit that file into your project. I personally still believe it has value in meson because it can serve as reference. You can generate it with meson initially then commit the file. It can also be useful for meson unit tests, many of them has some variant of this, it's boring to copy that file everywhere. I'm yet to see a single project who does that correctly, delegating that to meson means every dev does not have to know all the details, they can just trust meson.

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = '#pragma once\n'
```
I would rather not use `#pragma once` here. I think we should use include guards instead since they are much more universal.

xclaesse replied:
I never understood why people use header guards, as far as I know pragma once is universal and faster because it does not have to preprocess the whole file. It reads only the first line the first time it's included and next time it does not even open the file at all.

That said, I'm not against changing that here, should not make a real difference for a small file like that.

tristan957 replied:
No sense in changing it until there is more consensus on this PR

eli-schwartz replied:
Since we already produce pragma once for configure_file, it makes sense to produce it for anything else too as there is no advantage to doing so *inconsistently*.

xclaesse replied:
I didn't know there is precedent for using `#pragma once` in code generated by Meson. Since nobody ever complained, we can assume that's fine.

eli-schwartz replied:
I'm not certain whether anyone complained on the meson bugtracker.

Certainly, people complain on general C/C++ discussion sites that it isn't "portable" despite their refusal to use any compiler other than GCC, clang, or MSVC on account of the lot of the remainder being "half-baked garbage that doesn't implement standards from the last decade" and in some cases "is working towards c99 compliance but we're not quite there yet, please send patches".

Of course all 3 of those support pragma once.

Meanwhile, the standards committee doesn't want to standardize pragma once because they are concerned that the language specification shouldn't mandate that source code come inside files, because filesystems are mysterious beasts that are considered way too complicated for Serious Business™ like a programming language standard to depend on. What should the programming language describe as the correct way to handle headers that are a symlink? Or even worse, what about hardlinks??? No, no, this "filesystem" technology is too unpredictable.

xclaesse replied:
It's impossible to write C without hitting a million UB. At some point when all serious implementation does the same thing, who cares what the spec says...

eli-schwartz replied:
I find it hard to out into words, how utterly upset I was to (recently) discover that the spec says if you don't include an end of line character at the end of your file (the posix definition of a text, rather than binary file, is that it has posix line terminators)...

... Then because C/C++ source code is *text files*, and a file that isn't a posix text file is technically not C/C++ source code, the spec declares this, too, is UB. And not something sane like I don't know, an input processing "error: an input file was specified that isn't a source code file". Or even just saying "an implementation is permitted to accept inputs other than source files as an implementation extension, as long as it successfully parses as valid C/C++".

But no, they outright declare that compilers have a full license to generate UB and make demons fly out of your nose instead of doing what your code's C/C++ mapping says.

Nasal demons for your line endings. It's obscene. But they don't want to standardize pragma once ***because filesystems***.

tristan957 commented on mesonbuild/interpreter/compiler.py:
```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
+                ''')
+        content += textwrap.dedent(f'''
+            #if (defined(_WIN32) || defined(__CYGWIN__)) && !defined({static_compilation})
+            #  define _{api}_EXPORT __declspec(dllexport)
```
Why are these prefixed with underscore? Underscore-prefixed macros are reserved for the compiler

dnicolodi replied:
I though that the double leading underscore is reserved, not the single one.

xclaesse replied:
I think that as well, but if in doubt I can use triple underscore lol

tristan957 replied:
https://stackoverflow.com/a/17307796/7572728 has some good info

eli-schwartz replied:
Why do they even need to be private? :D

xclaesse replied:
> Why do they even need to be private? :D

Fair enough, removed the underscore. There is even a niche use-case for using `MYPROJECT_API_EXPORT`macro, glib does it for functions declarations that are exposed by modules: https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/giomodule.h#L130.

xclaesse commented:
Updated this PR for 1.3.0. AFAIK there is no strong opposition, just slight concern that it might not be as useful as it seems?

eli-schwartz commented:
> Updated this PR for 1.3.0. AFAIK there is no strong opposition, just slight concern that it might not be as useful as it seems?

The objection was more than slight concern, but it was also about the belief that this is badly suited to being a compiler attribute and better suited to being e.g. a module named something like "codesnippets".

xclaesse commented:
@eli-schwartz what other method would you see in that module? I never saw a concrete proposal, and making a module for a single method seems overkill. IMHO a compiler method is by far the most intuitive place, I really don't why that's a problem.

dnicolodi commented:
> making a module for a single method seems overkill

Well, technically, the `python` module exposes only one function... 😄

dnicolodi commented on docs/markdown/snippets/symbol_visibility_header.md:
```diff
@@ -0,0 +1,10 @@
+`__declspec(dllexport)`, `__declspec(dllimport)` or
+`__attribute__((visibility("default")))`.
+
+Meson can know generate a header file that defines exactly what's needed for
```
`s/know/now/`

xclaesse commented on docs/yaml/objects/compiler.yaml:
```diff
@@ -659,3 +659,105 @@ methods:
+        generated header.
+  description: |
+    Generate a header file that defines a macro to be used to mark all public APIs
+    of a library. Depending on the type of compiler and the platform, this will
```
> Typically this would be `__declspec(dllexport)`, `__declspec(dllimport)` for MSVC

I think that's a common misconception to think this is a MSVC thing. This is a **Windows** thing, even with GCC. At least as far as I understood.

https://gcc.gnu.org/onlinedocs/gcc/Microsoft-Windows-Function-Attributes.html
> On systems that support the visibility attribute, this attribute also implies “default” visibility.

It implies default visibility, but is not synonym to it, that's what I understand from that doc at least.

Removed `the type of compiler and` in my sentence, because it depends more on the platform than compiler.

dnicolodi replied:
I indeed thought so, but your sentence indeed cause me to doubt my recollections. I do almost no development on Windows. However, if the header does not depend on the compiler, someone could ask why this is a compiler object method, and we are back to the discussion about where this facility belongs...

The new sentence is clearer, but still gives the impression that different headers are generated on different platforms. I think it is important to communicate that the generated header is static. That's what my reformulation was trying to achieve.

xclaesse replied:
> think it is important to communicate that the generated header is static.

Added a sentence saying that explicitly.

> However, if the header does not depend on the compiler, someone could ask why this is a compiler object method, and we are back to the discussion about where this facility belongs...

It does not have to be a compiler method, but I personally think it's the most natural place. OTOH, it's not a hill I'm ready to die on, if @eli-schwartz really wants a module just for that, let's do it. Could be `meson.symbol_visibility_header()` too. But it's not clear he's going to accept it in that form, feels like he's going to reject it in any shape of form, but I could be wrong.

eli-schwartz replied:
I said I'd be happy with it as a module, and I mean it. :)

I do think that a module makes some sense, because there's other forms of snippets we may very well want too. Another possible use case is to generate test code snippets for some of the fancier autotools macros that projects sometimes find themselves using in order to test which of several different API conventions are used for the same meta-concept. Stuff like PAM_STRERROR_TWO_ARGS.

Why do you think that only visibility headers are going to end up being widely used/needed? :)

(This is also why I think the `meson` builtin object isn't any better of a location than the compiler object.)

bugaevc replied:
Just a drive-by note, potentially languages other than C may need visibility headers, which is why it makes sense to put it into a C compiler method.

xclaesse replied:
> Why do you think that only visibility headers are going to end up being widely used/needed? :)

Because I haven't seen requests for anything else yet, but I might very well just have missed them.

> potentially languages other than C may need visibility headers

Any specific language in mind? Currently this PR covers C/C++ and ObjC/ObjC++ which all use the same header file.

xclaesse commented:
Updated this PR to add `snippets` module.

tristan957 commented on mesonbuild/modules/snippets.py:
```diff
@@ -0,0 +1,98 @@
+            'symbol_visibility_header': self.symbol_visibility_header_method,
+        })
+
+    @typed_kwargs('compiler.symbol_visibility_header',
```
Looks like this didn't get renamed

xclaesse replied:
done

tristan957 commented on test cases/common/262 symbol visibility header/subprojects/sub/meson.build:
Maybe the tests should move into a snippets module subfolder?

xclaesse replied:
good point, done

tristan957 commented:
May want to change the PR title too!

ePirat commented:
> Should we add a license header to the generated header? Something like Unlicense or CC0? Or add a preamble/prefix kwarg of some sort?

IMO the best solution to this and the `#pragma` vs classic include guards issue is to add the possibility to allow to just get the string so projects and configure_file their own header with their own include guards, license and whatever else they might want to add.

eli-schwartz commented:
We don't add a license for generated config.h files because they're just a list of constants that have been format-shifted into C preprocessor statements. I can see the argument that this produced header is a bit more, well, creative work.

What's the purpose of the license in this case, then? Do we want to explicitly disavow any copyright claim against people using the produced header? Probably, that's the usual reason to go with a license that claims to be equivalent to a public domain dedication.

The frustrating thing is that there's a lot of disagreement about what kinds of public domain dedication "count". Various groups will insist that CC0 isn't free enough, or that unlicense isn't "clear", etc. For example, Google says its employees cannot contribute to projects under either of those two licenses as part of its internal policy guidance on:

> Forbidden patches
> 
> Our general philosophy is that we do not allow patches to projects that Google cannot use, either because of the terms of the license or Google policy.

xclaesse commented:
AFAIK if we want a license it should be Meson's Apache 2.0, because that's basically copying a string from Meson source into a file. I wouldn't mind adding `/* SPDX-License-Identifier: Apache-2.0 */` line on top.

But then raise the question of where did I found that code snippet... it is _inspired_ from what I saw in glib/gstreamer (LGPL2+), but not exactly (because I had to fix them in the process). It is also _inspired_ from GCC documentation https://gcc.gnu.org/wiki/Visibility (GPL?).

ePirat commented:
> AFAIK if we want a license it should be Meson's Apache 2.0, because that's basically copying a string from Meson source into a file. I wouldn't mind adding `/* SPDX-License-Identifier: Apache-2.0 */` line on top.

This will be unusable at least in VLC then, as it would force the rest of the code to (L)GPL 3 to be compatible and I am not even sure if that compatibility extends to source-level and not just linking…

xclaesse commented:
Let's just do nothing and everyone will be happy. As the author of that snippet I hereby claim that everyone can do what the fuck they want with it:  `SPDX-License-Identifier: WTFPL` :D

xclaesse commented:
More seriously, I think it's fine for GPL3: https://en.wikipedia.org/wiki/Apache_License#Compatibility

eli-schwartz commented:
> Let's just do nothing and everyone will be happy. As the author of that snippet I hereby claim that everyone can do what the fuck they want with it: `SPDX-License-Identifier: WTFPL` :D

That too is on the Google "do not use" list. :P My understanding is that Google is mad that it isn't actually Public Domain, so the code is still provided under a "license", but the license says you can "*do* whatever you want" but doesn't say you can "*have* a warranty disclaimer". :facepalm:

Their documentation on it appears to have vanished but [the Internet Archive](https://web.archive.org/web/20181022125422/https://opensource.google.com/docs/thirdparty/licenses/#wtfpl-not-allowed) does not forget:
> Code released under the [WTFPL](http://www.wtfpl.net/) cannot be used at Google. This license has a large number of issues (lack of warranty disclaimer, very vague rights grant), and was also rejected as an open source license by OSI. We also do not allow contribution to projects under the WTFPL.

Perhaps its disappearance is a sign that it became acceptable to use, but I doubt it since it is still listed as one of the licenses that Googlers aren't allowed to contribute to: https://opensource.google/documentation/reference/patching#forbidden

BSD0 is however on the approved list as an explicit exception to "we at Google do not like Public Domain dedications", and is commonly considered to be "not actually a Public Domain dedication but effectively equivalent".

eli-schwartz commented:
> Let's just do nothing and everyone will be happy.

And in particular this: I am not overly worried about whether or not people can freely use a function to generate code snippets, but "do nothing and everyone will be happy" is an *incredibly* humorous thought, because this is a topic that people make a very good livelihood off of convincing the entire tech world to be deeply unhappy about it.

xclaesse commented:
Anyway, let me know what you want, I couldn't care less. I'm not a lawyer.

tristan957 commented:
> IMO the best solution to this and the `#pragma` vs classic include guards issue is to add the possibility to allow to just get the string so projects can configure_file their own header with their own include guards, license and whatever else they might want to add.

I really like this idea. @xclaesse what if we made this more similar to vcs_tag()?

xclaesse commented:
Returning the string to be able to add it into an existing `configure_file()` would make sense, that's the kind of things many project already have in their `config.h`. IMHO that would require to add at least `fs.write()` method to be able to write that string verbatim into a new file, and also have proper solution for https://github.com/mesonbuild/meson/issues/6259.

I don't think that solves the license question, the problem is exactly the same, are you allowed to paste that code into a file with your project license. But unless one of us is secretly a lawyer specialized on that matter, I think we are just talking without knowing. If a big company wants to pay a few millions their legal team to debate on this we'll probably have a answer in a decade or two.

tristan957 commented:
What I am thinking is more along the lines of:

```c
// header.h.in

// preable...

@SYMBOL_VISIBILITY_MACROS@

// function decls, etc.
```

```meson
header = snippets.symbole_visibility_macros(input: 'header.h.in', output: 'header.h')
```

xclaesse commented:
That makes API more complicated to use for no real benefit IMHO. Hooking it with configure_file() does have the benefit to make the snippet part of an already existing file in the project instead of adding an extra one.

tristan957 commented:
The benefit I see is that in the current approach all projects now have to install a `visibility.h` into the includedir. While the proposed approach allows them to embed in any file they want.

xclaesse commented:
Sure, but then doing it through `configure_file()` seems much more flexible because you can substitute other keys in a single step.

tristan957 commented:
> Sure, but then doing it through `configure_file()` seems much more flexible because you can substitute other keys in a single step.

Right. This is similar the vcs_tag() idea I talked about on Matrix

xclaesse commented:
@tristan957 I updated this PR, and added `symbol_visibility_string()` that returns the string that can be used with `configure_file()` instead of writing a file.

xclaesse commented on docs/markdown/Snippets-module.md:
```diff
@@ -0,0 +1,118 @@
+  It will be converted to upper case with all non-alphanumeric characters replaced
+  by an underscore `_`. It is only used for `api`, `compilation` and
+  `static_compilation` default values.
+- `api`: Name of the macro used to mark public APIs. Defaults to `<NAMESPACE>_API`.
+- `compilation`: Name of a macro defined only when compiling the library.
+   Defaults to `<NAMESPACE>_COMPILATION`.
+- `static_compilation`: Name of a macro defined only when compiling or using
+  static library. Defaults to `<NAMESPACE>_STATIC_COMPILATION`.
```
It all has default value, the flexibility is intended mostly for wrapdb ports where you don't control how the project names those macros.

xclaesse replied:
Typically cmake projects using [GenerateExportHeader](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html).

bonzini commented on mesonbuild/modules/snippets.py:
```diff
@@ -0,0 +1,112 @@
+                  KwargInfo('static_compilation', (str, NoneType)),
+                  KwargInfo('static_only', bool, default=False))
+    @noPosargs
+    def symbol_visibility_string_method(self, state: ModuleState, args: T.Tuple[str], kwargs: 'SymbolVisibilityHeaderKW') -> str:
```
I am just worried that this becomes a dumping ground with no rhythm and rhyme. Sure `c` is not hundred percent accurate but it's better than nothing. I will let others judge the color that they think is best for the bike shed.

xclaesse commented on docs/markdown/Snippets-module.md:
```diff
@@ -0,0 +1,118 @@
+apiheader = snippets.symbol_visibility_header('apiconfig.h',
+  static_only: default_library == 'static')
+install_headers(apiheader, 'lib.h', subdir: 'mylib')
+lib = library('mylib', 'lib.c',
```
It's generated at configure time, so it's not needed. You add generated headers in sources when it's generated at build-time, so meson needs to know the build order.

xclaesse commented on docs/markdown/Snippets-module.md:
```diff
@@ -0,0 +1,118 @@
+of a library. Depending on the platform, this will typically use
+`__declspec(dllexport)`, `__declspec(dllimport)` or
+`__attribute__((visibility("default")))`. It is compatible with C, C++,
+OBJC and OBJC++ languages. The content of the header is static regardless
```
fixed

tristan957 reviewed (commented):
This looks awesome. Thanks for sticking with it!
